docs/specs/ML.Net Sweepable API.md

```diff
@@ -0,0 +1,162 @@
+# AutoML.Net Sweepable API proposal

```

Comment:
@jwood803  ( I saw you're like ) we'd love to hear your feedback as well.

Reply:
Hey, @JakeRadMSFT. I definitely like having this API! This helps make it more on part with scikit-learn that has grid search. Plus, I believe this will help model creators get better models.

Reply:
> I'm open to keeping these layers internal if we need to

My suggestion would be to start with them internal, and ensure it can meet the existing AutoML scenarios. Then make them public as we go when we have data that supports they are needed to be public.

> The end goal is to only have one ML.NET AutoML.

💯

Reply:
Sounds good, if we want to start from internal first, a smooth start without breaking change can be search space, either through package reference or source code. The reasons are
- search space has the least dependency. It only depends on newtonsoft, which makes it easy to introduce.
- after introducing, we can use it to create search space for existing trainers in AutoML.Net. Which provides us a chance to improve the performance of AutoML.Net experiments by using larger, better search space in hpo as well.

Reply:
Sounds like a fine plan, including keeping as internal for a bit.

If it fits within your methods, I'd recommend having the _default_ sweeping space { range, scaling, data type } as part of the component being swept (as [current](https://github.com/dotnet/machinelearning/blob/510f0112d4fbb4d3ee233b9ca95c83fae1f9da91/src/Microsoft.ML.FastTree/FastTreeArguments.cs#L508-L509) params). For AutoML․NET, due to non-great reasons, we kept a duplicate copy. Traveling with the component is more clean since the ranges show up (and disappear) with the availability of the component. For instance, as the user includes the NuGet for FastTree, the ranges are immediately available. In addition the ranges can be created inline with the new component. 

Some historical perspective, there is also a command-line sweeper in ML․NET, which may be working (anything untested is assumed non-functional). This does let users sweep an individual (or sets of) components including fully switching out portions of the pipeline. The MAML command-line sweeper was very powerful, but hard to use. Mentioned previously: https://github.com/dotnet/machinelearning/pull/5019#issuecomment-817105977

Reply:
> It only depends on newtonsoft

If possible, can we use System.Text.Json?

Reply:
> If possible, can we use System.Text.Json?

Currently no because auto-binding experience depends heavily on `JToken`. However, there's a plan to remove this dependency so in the future maybe.

Reply:
### Resolved